### ISSUE TITLE
TF 2.0:  tf.GradientTape().gradient() returns None

### ISSUE THREAD

**Issue by sailormoon2016** [tensorflow/tensorflow#30190]
I designed my own loss function for my graduate research, it calculates the distance between the histogram of each loss in a batch and normal distribution. I am implementing this loss function in the setting of Tensorflow 2.0 [tutorial](https://www.tensorflow.org/beta/tutorials/eager/custom_training_walkthrough) about Iris flower classification.

I checked my loss value and type, they are same as the one in tutorial, but the `grads` from my `tape.gradient()` is `None`.

This is done in Google Colab with:

`TensorFlow version: 2.0.0-beta1`

`Eager execution: True`

My code block for loss and gradient:
```
def loss(model, x, y):
  y_ = model(x) # y_.shape is (batch_size, 3)
  losses = []
  for i in range(y.shape[0]):
    loss = loss_object(y_true=y[i], y_pred=y_[i])
    losses.append(float(loss))
  dis = get_distance_between_samples_and_distribution(losses, if_plot = 0)
  return tf.convert_to_tensor(dis, dtype=np.float32)

def grad(model, inputs, targets):
  with tf.GradientTape() as tape:
    loss_value = loss(model, inputs, targets)
    tape.watch(model.trainable_variables)
  return loss_value, tape.gradient(loss_value, model.trainable_variables)

loss_value, grads = grad(model, features, labels)
print("loss_value:",loss_value)
print("type(loss_value):", type(loss_value))
print("grads:", grads)
################################################# Output:
loss_value: tf.Tensor(0.21066944, shape=(), dtype=float32)
type(loss_value): <class 'tensorflow.python.framework.ops.EagerTensor'>
grads: [None, None, None, None, None, None]

```
The code in tutorial is:
```
loss_object = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

def loss(model, x, y):
  y_ = model(x)
  return loss_object(y_true=y, y_pred=y_)

def grad(model, inputs, targets):
  with tf.GradientTape() as tape:
    loss_value = loss(model, inputs, targets)
    tape.watch(model.trainable_variables)
  return loss_value, tape.gradient(loss_value, model.trainable_variables)

loss_value, grads = grad(model, features, labels)
print("loss_value:",loss_value)
print("type(loss_value):", type(loss_value))
print("grads:", grads)
################################################# Output:
loss_value: tf.Tensor(0.56536925, shape=(), dtype=float32)
type(loss_value): <class 'tensorflow.python.framework.ops.EagerTensor'>
grads: [<tf.Tensor: id=9962, shape=(4, 10), dtype=float32, numpy=
array([[ 0.0000000e+00,  6.5984917e-01,  3.0700830e-01, -7.5234145e-01,
      ......
```
I feel the calculation of my self defined loss should not matter since the data type and shape are same, but in case it does, here is my loss function:
```
def get_distance_between_samples_and_distribution(errors, if_plot = 1, n_bins = 5):
  def get_middle(x):
    xMid = np.zeros(x.shape[0]//2)
    for i in range(xMid.shape[0]):
      xMid[i] = 0.5*(x[2*i]+x[2*i+1])
    return xMid

  bins, edges = np.histogram(errors, n_bins, normed=1)
  left,right = edges[:-1],edges[1:]
  X = np.array([left,right]).T.flatten()
  Y = np.array([bins,bins]).T.flatten()
  X_middle = get_middle(X)
  Y_middle = get_middle(Y)
  distance = []
  for i in range(X_middle.shape[0]):
    dis = np.abs(scipy.stats.norm.pdf(X_middle[i])- Y_middle[i])
    distance.append(dis)
  distance2 = np.power(distance, 2)
  
  return sum(distance2)/len(distance2)
```
**I searched and tried adding `tape.watch()`, checking the indentation of my return as suggested in [issue #29942](https://github.com/tensorflow/tensorflow/issues/29942) but they did not fix this `None` problem. I will very appreciate any suggestion for fixing this. Thanks!**

The definition of class `tf.GradientTape` is [here](https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/python/eager/backprop.py)

**Comment by sailormoon2016** [comment]
it's solved

**Comment by jvishnuvardhan** [comment]
@sailormoon2016 If possible post how you resolved so that it will help the community. Thanks!

**Comment by sailormoon2016** [comment]
I figured my self defined loss function may not be differentiable, so I modified my loss function to make sure the calculation is differentiable, then the results come out as expected. My customized loss function did give a better accuracy compared to the default cross-entropy. you can find my code [here](https://github.com/sailormoon2016/Transportation/blob/master/ErrorShaping/tf2_flower.ipynb) if you are interested in. Let me know if you have more question.

**Comment by jvishnuvardhan** [comment]
@sailormoon2016 Thanks for the information and code. It will help others in the community. Greatly appreciated. Thanks!

**Comment by shubham-malaviya** [comment]
Below link is helpful in finding problems related to GradientTape :

[Tensorflow 2.0 Pitfalls](http://blog.ai.ovgu.de/posts/jens/2019/001_tf20_pitfalls/index.html)

**Comment by sharkdeng** [comment]
> it's solved

Can you share how you solved this?

**Comment by JHUB-JNTUH** [comment]
I have a stranger loss function where I am converting the output to embeddings and calculating distance of embeddings. This is not at all differentiable even if embeddings are built with TF math ops. Is there an example for writing my own differentiation code as this is not realisable through TF automatic differentiation?

**Comment by SandeepNaidu** [comment]
I have a stranger loss function where I am converting the output to embeddings and calculating distance of embeddings. This is not at all differentiable even if embeddings are built with TF math ops. Is there an example for writing my own differentiation code as this is not realisable through TF automatic differentiation?

**Comment by mitran27** [comment]
Hi I had the same issue but after indenting loss calculation within tf.GradientTape it started to work

**Comment by varungupta31** [comment]
Requesting help with my Issue #56657.
Kindly help, if stil in-touch with GradCam.

Thank you :)
